en/21/06.md

```diff
@@ -0,0 +1,199 @@
+---
+title: Function Declarations
+actions: ['checkAnswer', 'hints']
+material:
+  editor:
+    language: rust
+    startingCode:
+      "lib.rs": |
+        #![no_std]
+
+        multiversx_sc::imports!();
+        multiversx_sc::derive_imports!();
+
+        #[derive(NestedEncode, NestedDecode, TopEncode, TopDecode, TypeAbi)]
+        pub struct Zombie<M: ManagedTypeApi> {
+            name: ManagedBuffer<M>,
+            dna: u64,
+        }
+
+        #[multiversx_sc::contract]
+        pub trait ZombiesContract {
+
+          #[init]
+          fn init(&self) {
+            self.dna_digits().set(16u8);
+            self.zombies_count().set(1usize);
+          }
+
+          // start here
+
+          #[storage_mapper("dna_digits")]
+          fn dna_digits(&self) -> SingleValueMapper<u8>;
+
+          #[storage_mapper("zombies_count")]
+          fn zombies_count(&self) -> SingleValueMapper<usize>;
+
+          #[storage_mapper("zombies")]
+          fn zombies(&self, id: usize) -> SingleValueMapper<Zombie<Self::Api>>;
+        }
+    answer: |
+      #![no_std]
+
+      multiversx_sc::imports!();
+      multiversx_sc::derive_imports!();
+
+      #[derive(NestedEncode, NestedDecode, TopEncode, TopDecode, TypeAbi)]
+      pub struct Zombie<M: ManagedTypeApi> {
+          name: ManagedBuffer<M>,
+          dna: u64,
+      }
+
+      #[multiversx_sc::contract]
+      pub trait ZombiesContract {
+
+        #[init]
+        fn init(&self) {
+          self.dna_digits().set(16u8);
+          self.zombies_count().set(1usize);
+        }
+
+        fn create_zombie(&self, name: ManagedBuffer, dna: u64) {
+            self.zombies_count().update(|id| {
+              self.zombies(id).set(Zombie { name, dna });
+              *id +=1;
+            });
+        }
+
+        #[storage_mapper("dna_digits")]
+        fn dna_digits(&self) -> SingleValueMapper<u8>;
+        
+        #[storage_mapper("zombies_count")]
+        fn zombies_count(&self) -> SingleValueMapper<usize>;
+
+        #[storage_mapper("zombies")]
+        fn zombies(&self, id: &usize) -> SingleValueMapper<Zombie<Self::Api>>;
+      }
+---
+
+## Functions in Rust
+
+A function declaration in Rust looks like the following:
+
+```
+pub fn eat_hamburgers(&self, amount: u32) {
+
+}
+```
+
+This is a function named `eat_hamburgers` that takes 2 parameters: a reference to `self` and a `u32`. For now the body of the function is empty. Note that we're specifying the function visibility as `public`. Without the `pub` keyword the function would be private, visible only within the entity that implements it.
+
+What is a reference type you ask?
+
+Well, there are three ways in which you can pass an argument to a Rust function:
+
+ * By value, which means that the ownership over the parameter's value is passed to your function. This blocks the use of the initial parameter once the function call had ended since having it consummed by the function.
```

Comment:
Are you sure this is how it works? I have a background in C++, and I find it confusing.

Reply:
Ownership in rust is quite different than the one of C / C++.

Rust doesn't make local copies of the variable passed but rather passing the ownership of it. This is why if you want to change it and have access to it further ourside yout function you should use a reference, or a mutable reference if you also want to change it inside.

I will explain a bit more clearly